plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -432,28 +432,49 @@ spendBudgetCek = let (CekBudgetSpender spend) = ?cekBudgetSpender in spend
 -- see Note [Scoping].
 -- | Instantiate all the free variables of a term by looking them up in an environment.
 -- Mutually recursive with dischargeCekVal.
-dischargeCekValEnv :: CekValEnv uni fun -> Term Name uni fun () -> Term Name uni fun ()
-dischargeCekValEnv !valEnv =
-    -- We recursively discharge the environments of Cek values, but we will gradually end up doing
-    -- this to terms which have no free variables remaining, at which point we won't call this
-    -- substitution function any more and so we will terminate.
-    termSubstFreeNames $ \name -> do
-        val <- lookupName name valEnv
-        Just $ dischargeCekValue val
+dischargeCekValEnv :: forall uni fun. CekValEnv uni fun -> Term NamedDeBruijn uni fun () -> Term NamedDeBruijn uni fun ()
+dischargeCekValEnv valEnv = go 0
+ where
+  -- an extra extended level (extLvl) keeps track of how far (deep) we have lambda-descended.
+  go :: Word -> Term NamedDeBruijn uni fun () -> Term NamedDeBruijn uni fun ()
+  go !extLvl =  \case
+    -- increase the extLvl
+    LamAbs ann name body -> LamAbs ann name $ go (extLvl+1) body
+    var@(Var _ (NamedDeBruijn _ ix)) -> let n = coerce ix :: Word  in
+        if extLvl < n
+        then
+            case (valEnv `addLvl` extLvl) `Env.index` n of
+                Just val -> dischargeCekValue val
+                -- FIXME: proper error handling of ill-scoped input, as the previous MalformedScript error did.
+                Nothing  -> error "ill-scoped input"
```

Comment:
I was just thinking if we should do it properly, using the MonadError of CekM

Reply:
> I think maybe just leave it in this case, it corresponds to a free variable

There's an error for that, see `OpenTermEvaluatedMachineError` in `lookupVarName`.

Reply:
The thing is, we often call this when we're *already* reporting an error. So I think we might rather report errors with free variables in them. For example:

```
{ (\x -> y) (con integer) }
```
will fail because you're instantiating something that's not an abstraction, and we'll never evaluate inside the lambda to see the body has a free variable. It would be annoying to then fail when creating the error, nicer to just report the term with free variables (I think?).